### ISSUE TITLE
Remove direct DOM objects access trough DomEventData

### ISSUE THREAD

**Issue by szymonkups** [ckeditor/ckeditor5-engine#651]
`DomEventData` exposes native DOM objects as its properties: `domEvent` and `domTarget`. `DomEventData` is an abstraction over DOM event system and should not allow to directly use these objects.

There are situations when we would like to mimic observers' events but we cannot do this unless we use native DOM event objects. 

I've also noticed that `SelectionObserver` fires event containing native selection object. Is it necessary? https://github.com/ckeditor/ckeditor5-engine/blob/master/src/view/observer/selectionobserver.js#L153

**Comment by Reinmar** [comment]
We'll need to check whether we don't depend on the DOM stuff already. Also, I'm a bit worried that hiding `domEvent` will render the event very limited because other developers may want to access things that we haven't yet exposed (and there are maaany DOM evt properties out there). Every time they will want something from the event they will have to come to us. So I'd consider keeping the `domEvent` but as a private property, so it can be accessed as a last resort. The documentation can also explain this ("you can access _domEvent but please report a ticket what do you miss in the facade class").

**Comment by pjasiun** [comment]
I disagree that keeping `_domEvent` is a good idea. If we will fire custom events, without `_domEvent` they may crash in all cases when a developer will do something based on this missing event. Most probably he will not know that this event might be missing. It's better to have them asking us for the missing API because then we will be able to figure out how custom events should behave in these cases. Think about the clipboard event. If we let users take DOM event, native data transfer and create plugins based on this event, we will have a problem soon, if we decide to add paste bin.

**Comment by Reinmar** [comment]
> It's better to have them asking us for the missing API because then we will be able to figure out how custom events should behave in these cases. Think about the clipboard event. If we let users take DOM event, native data transfer and create plugins based on this event, we will have a problem soon, if we decide to add paste bin.

This way of thinking will make developers hate you. How quickly can you add the missing properties? In 2-3 days? In the next release (one month)? It's still at least couple of days too late. It's absolutely unrealistic to expect that someone will care about your perfect architecture if it causes so much trouble. People will soon start avoiding the observers at all which will be an even bigger problem than what you mentioned with the tests.

BTW, your point about failing tests is also incorrect, because then you expect that people will actually write tests for their code. 90% won't and the other 10% will then perfectly see that they miss some property and this should raise the chance that such thing will ever get reported.

**Comment by Reinmar** [comment]
> Think about the clipboard event. If we let users take DOM event, native data transfer and create plugins based on this event, we will have a problem soon, if we decide to add paste bin

Again, wrong. If they'd be forced to use the native data transfer, then it means that we introduced absolutely broken event. I still choose that they are able to do something that telling them that there's no option to access the native data transfer and that they have to wait one month for a fix.

**Comment by pjasiun** [comment]
My point is that if they will use `_domEvent` they will be happy at the beginning, that they find this property, and very sad when they add a feature which fire event without `_domEvent` and found out that their code does not work.

**Comment by pjasiun** [comment]
On the other hand, I agree we could add `_domEvent` as a private property, so if one need, he can use it, but since it is private he should accept that it can change.
